### PR TITLE
chore: add data to Firebase 

### PR DESCRIPTION
**Pull request by BumbleFlash**
**Please do not merge. Work in Progress**

This PR is to write the code that pushes `Event `data to Firebase FireStore. 

TODO:
- [x] Add utility class to write Firebase events
- [x] Write play/pause event to Firebase
    - [x] Check song ending and next song playing (no user interaction, for play event only)
    - [x] Check back button to start playing song again (for play event only)
    - [x] UI action for notification interaction for play/pause
    - [x] Capture seekPosition of song

Future PRs will handle the user enrollment process, including generating the random user ID.

### CODE REVIEW

**Comment by barbeau** [comment]
@BumbleFlash Thanks for opening this! If this is still in a draft state, I'd suggest converting it to a "Draft" pull request to indicate this.

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    public static void saveEvent(Event event, String userId, String recordID){
+        DocumentReference documentReference =
+                getFirebaseDocReferenceByUserIDAndRecordId(userId, recordID,
+                        EventConstants.FIREBASE_EVENT_FOLDER);
+
+        documentReference.set(event)
+                .addOnCompleteListener(task -> {
+                      if (task.isSuccessful()){
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java]:
```suggestion
                      if (task.isSuccessful()) {
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    private static String buildDocumentPathByUid(String uid, String folder) {
+        return "users/" + uid + "/" + folder;
+    }
+
+    private static DocumentReference getFirebaseDocReferenceByUserIDAndRecordId(String userId,
+                                                                               String recordId,
+                                                                               String folder){
+        String path = buildDocumentPathByUid(userId, folder);
+        FirebaseFirestore firebaseFirestore =FirebaseFirestore.getInstance();
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java]:
```suggestion
        FirebaseFirestore firebaseFirestore = FirebaseFirestore.getInstance();
```

File: app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package edu.usf.sas.pal.muser.constants;
+
+public class EventConstants {
+
+    public static String FIREBASE_EVENT_FOLDER = "events";
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java]: Let's change the name of this to `player-events`. We may record other data like activity transitions in the future, so we'll want to be able to clearly distinguish what events are related to the music player and what come from another source. So we should have paths like:

```
/users/23jlfkdfd/player-events/sdfuidf
/users/23jlfkdfd/activity-transitions/klwjlgi
...
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java [hunk 4]
```diff
@@ -18,13 +17,14 @@
      * @param song - The song for which the event occurred.
      * @param capturedEvent - The event that was captured.
      * @param context - The context of the Fragment.
-     * @return - Event object
      */
-    public static Event newEvent(Song song, EventType capturedEvent, Context context){
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: Was this change intentional? Seems like this method won't have any effect now.

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+import android.util.Log;
+
+import com.google.firebase.firestore.DocumentReference;
+import com.google.firebase.firestore.FirebaseFirestore;
+
+import edu.usf.sas.pal.muser.constants.EventConstants;
+import edu.usf.sas.pal.muser.model.Event;
+
+public class EventFirebaseIOUtils {
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java]: Per my other comment, let's change the name of this class to the more generic `FirebaseIOUtils`:

```suggestion
public class FirebaseIOUtils {
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    private static final String TAG = "EventFirebaseIO";
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventFirebaseIOUtils.java]:
```suggestion
    private static final String TAG = "FirebaseIO";
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package edu.usf.sas.pal.muser.util;
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java:1]: Doesn't seem like this class is used currently?

File: app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java [hunk 8]
```diff
@@ -378,4 +382,12 @@ private void enableStrictMode() {
                 .penaltyFlashScreen()
                 .build());
     }
-}
+
+    public static ShuttleApplication get() {
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java:400]: Doesn't seem like these new methods and variables are used currently? Also, `mApp` is never assigned, so this will throw an NPE.

**Review by barbeau — CHANGES_REQUESTED** [review]
@BumbleFlash Thanks! Generally this is looking good, although it's hard to review a PR when it's not clear what is intended to be complete and what work still remains.

Looking at the Firebase impl in OBA, there is still a lot that needs to be ported over. So let's follow this process:
1. I've create a new branch `firebase-full` off of `usf`, and changed this PR to target the `firebase-full` branch. If there are smaller PRs with pieces of Firebase we can open those against the `firebase-full` branch and merge there until the Firebase implementation is complete, and then merge that full `firebase-full` branch into the `usf` branch.
1. I've added a checklist to your original comment (see https://docs.github.com/en/github/managing-your-work-on-github/about-task-lists). Please use these in future PRs so it's easier for me to know what you've implemented in a PR, and what you plan to still implement in the *same* PR before it gets merged to `firebase-full`. And if there are more pieces that need to be implemented in future PRs, please add a comment mentioning that too. I've taken a guess for this PR, but please change if you think differently.

So, please take a look at the OBA code and break up the Firebase impl into logical pieces, and we can do smaller PRs for as mentioned above for each of those pieces. It doesn't need to be a lot - if 2 PRs will do it (or even using this same 1 PR), that's fine. It will just make it easier for me to understand and review when it's clearer what this PR adds at the point of review, and the pieces that you still need to add.

**Comment by barbeau** [comment]
One more comment on process - when a PR is ready to be merged, it's status should be changed from "Draft".

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java (hunk 4 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: Correct - it doesn't have any effect right now. I wasn't sure if we'll have to use the return type anytime so I decided to make it `void` for now. We can always change it depending upon how we use the function later on.

File: app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java (hunk 7 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java:1]: Correct - It's not being used yet. I have written the functions to save Strings to the `SharedPreferences` and never actually called them anywhere.

File: app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java (hunk 8 above)
Comment (BumbleFlash) [reply app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java:400]: Before our discussion in Teams, I thought this was the part that was generating UserIDs. So I will have the code written the future commits for it. Also prior to our discussion earlier today, do you think we need to capture the APP_UIDs for Muser now?

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java (hunk 4 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: But we need this method to generate new events and populate the timestamp fields, right? I'd leave it returning `Event`, as I don't see how this works without it.

File: app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java (hunk 8 above)
Comment (barbeau) [reply app/src/main/java/com/simplecity/amp_library/ShuttleApplication.java:400]: No, we don't need the APP_UIDs in MUSER - that exists in OBA to generate a random ID on first execution and is sent with server API requests to get an idea of the number of clients server-side. We don't have an equivalent in MUSER, so we don't need it.

The UUIDs we do need are:
1. Firebase UUID for the user
1. Event UUIDs (generated from the Android UUID method)

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    public static void saveEvent(Event event, String userId, String recordID){
+        DocumentReference documentReference =
+                getFirebaseDocReferenceByUserIDAndRecordId(userId, recordID,
+                        EventConstants.FIREBASE_EVENT_FOLDER);
+
+        documentReference.set(event)
+                .addOnCompleteListener(task -> {
+                      if (task.isSuccessful()) {
+                          Log.d(TAG, "Event saved with ID " + documentReference.getId());
+                      }
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: `} else {` should all be on one line here.

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java (hunk 4 above)
Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: If we could call the function that saves data to firebase inside the `newEvent()`, then I think we won't be needing the return type. Can this be one of the right ways of going about it? Thank you.

Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: Hmmm, I see what you're saying. I think a cleaner separation would be to have a Firebase manager class be the one to save data to firebase, and keep this utility class simple as just generating new events (which would be called from the Firebase manager, possibly). I think this was the design of OBA, IIRC (not looking at the code now)?

Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java:23]: Yes, you're right - The `saveTravelBehaviour` function takes in a pojo object and I guess I'll follow the same flow to make it look more cleaner.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerFragment.java [hunk 10]
```diff
@@ -228,13 +229,18 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
                 Song song = getSong();
-                EventType eventType;
-                if (isPlaying())
-                    eventType = EventType.PLAY_MANUAL;
-                else
-                    eventType = EventType.PAUSE_MANUAL;
-                Log.d(TAG, "onViewCreated: " + eventType);
-                EventUtils.newEvent(song, eventType, getContext());
+                PlayerEventType playerEventType;
+                UiActionType uiActionType;
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerFragment.java]: Let's call these `UiEventType` if both this and `PlayerEventType` are going to implement a common `Event` interface.

File: app/src/main/java/edu/usf/sas/pal/muser/model/UiEvent.kt [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+/**
+ * [uiEventType] - The enum object for the type of action performed.
+ */
+(val uiEventType: UiEventType,
+ override val currentTimeMs: Long,
+ override val nanoTime: Long,
+ override val startTime: Long,
+ override val elapsedTime: Long,
+ override val song: SongData
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/UiEvent.kt]: `song` needs to be nullable here in case we have UI events that don't include a song

File: app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt [hunk 12]
```diff
@@ -1,24 +1,23 @@
...
- val startTime: Long,
- val elapsedTime: Long,
- val song: SongData
-)
\ No newline at end of file
+{
+    val currentTimeMs: Long
+    val nanoTime: Long
+    val startTime: Long
+    val elapsedTime: Long
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: I'm still not convinced that the current `elapsedTime` gets us what we need in terms of recording where in the track the user started and stopped it, as the current value looks like a clock time to me (e.g., "Aug 26 3:00pm"). 

So if you play the track at Aug 26 3:00pm, and then pause it at 3:02pm, and then resume playing it at 3:05pm, the `elapsedTime` value is 3:05pm (?). But that doesn't tell us that the user resumed playing the track 2 minutes into the track. This value is also redundant to `currentTimeMs`, which we can use to look at each event and calculate the duration between events.

I think we need more of a `trackPositionSec` that tells us how many seconds into the track an action occurred, like "120 seconds". So we know that a user played the track and started it at 120 seconds in, and then paused it again at 130 seconds in.

**Comment by barbeau** [comment]
A few quick notes after looking at Firebase data, which looks good!

For the Song data:
- [x] What's the units for `duration`? Let's change the field name to include it, like `durationSec`, and add it to the Javadocs too
- [x] What's the unit for `lastPlayed`? Similar on name change, and I believe it needs Javadocs too.

File: app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt (hunk 12 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Similar for `startTime` - how is that different from `currentTimeMs`?

Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Yes - You're right. the `elapsedTime` is just getting the current time and doesn't really work as we think it should. I guess we'll have to calculate and store `trackPositionSec` as you suggested.

Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Also, `startTime` updates and stores the timestamp when a track is played (not paused) by the user. I guess the attribute helps to keep track of when a particular song was last played?

**Comment by BumbleFlash** [comment]
> * [ ]  What's the units for `duration`? Let's change the field name to include it, like `durationSec`, and add it to the Javadocs too

Currently, the units for the `duration` is in milliseconds. 

> * [ ]  What's the unit for lastPlayed? Similar on name change, and I believe it needs Javadocs too.

I am still not sure what `lastplayed` is actually doing, Logging the value stored in the variable always gives 0, so it doesn't really help in deciphering its purpose.

File: app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt (hunk 12 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Can we get rid of `elapsedTime` and `startTime`? For our work, if we're tracking the `currentTimeMs` and `nanoTime` for each event, I don't see how these other variables add any additional info.

Comment (BumbleFlash) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Yes - I totally agree. Will remove them in the next commit.

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+                    if (task.isSuccessful()) {
+                        Log.d(TAG, "UiEvent saved with ID " + documentReference.getId());
+                    } else {
+                        logErrorMessage(task.getException(), "UiEvent Save Failed");
+                    }
+                });
+    }
+
+    private static String getRecordID(){
+        long rPrefix = PreferenceUtils.getLong(EventConstants.RECORD_ID, 0);
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java:67]: Per discussion today, we need to handle multithreading issues with incrementing and storing the integer counter.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerFragment.java [hunk 14]
```diff
@@ -228,13 +232,20 @@ public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceStat
...
                 presenter.togglePlayback();
                 Song song = getSong();
-                EventType eventType;
-                if (isPlaying())
-                    eventType = EventType.PLAY_MANUAL;
-                else
-                    eventType = EventType.PAUSE_MANUAL;
-                Log.d(TAG, "onViewCreated: " + eventType);
-                EventUtils.newEvent(song, eventType, getContext());
+                PlayerEventType playerEventType;
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/nowplaying/PlayerFragment.java]: Per discussion today, let's move player events to the `PlaybackManager.play()` and related methods for individual events, or as close to `notifyChange(InternalIntents.*_STATE_CHANGED);` as possible, to make sure the event we're tracking actually happened.

File: app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt (hunk 12 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: Per discussion today, let's change `elapsedTime` to `seekPositionMs` and get ride of `startTime`.

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java (hunk 13 above)
Comment (barbeau) [reply app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java:67]: This should be as easy as adding the `synchronized` keyword to the method signature.

File: app/src/main/java/com/simplecity/amp_library/playback/MusicService.java [hunk 15]
```diff
@@ -411,12 +416,12 @@ public int onStartCommand(Intent intent, int flags, int startId) {
                         if (isPlaying()) {
 
                             // It's not possible to be playing and the service not be started. No ANR
-
+                            newUiEvent(queueManager.getCurrentSong(), UiEventType.PLAY);
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/playback/MusicService.java]: Should this be `UiEventType.PAUSE`?

Also do you know if the above comment suggests that execution should never reach this state? That's how I read it without looking at the following code execution.

File: app/src/main/java/com/simplecity/amp_library/playback/PlaybackManager.java [hunk 16]
```diff
@@ -360,6 +367,7 @@ void setNextTrack() {
                 && !queueManager.getCurrentPlaylist().isEmpty()
                 && queueManager.nextPlayPos < queueManager.getCurrentPlaylist().size()) {
             final Song nextSong = queueManager.getCurrentPlaylist().get(queueManager.nextPlayPos).getSong();
+            // adds to queue, log here.
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/playback/PlaybackManager.java]: Is this a `// TODO` for a future player event? If so, could you label it with `// TODO` and a place holder even type so it's a bit clearer? Like:

`// TODO - this adds a song to the queue. Log as new PlayerEventType like PlayerEventType.ADDED_TO_QUEUE`

File: app/src/main/java/com/simplecity/amp_library/playback/PlaybackManager.java [hunk 17]
```diff
@@ -530,10 +540,15 @@ public void play() {
                 playAutoShuffleList();
             }
         }
-
+        Song song = queueManager.getCurrentSong();
+        newPlayerEvent(song, PlayerEventType.PLAY);
         notifyChange(InternalIntents.PLAY_STATE_CHANGED);
     }
 
+    static void newPlayerEvent(Song song, PlayerEventType playerEventType){
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/playback/PlaybackManager.java]: I don't think this needs to be `static`?

File: app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java [hunk 18]
```diff
@@ -135,11 +141,16 @@ public void load(@NonNull List<Song> songs, final int position, @NonNull UnsafeA
         }
 
         openCurrentAndNext.run();
-
+        newPlayerEvent(getCurrentSong());
         notifyMetaChanged();
         notifyQueueChanged();
     }
 
+    static void newPlayerEvent(Song song){
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java]: Same here - I believe this shouldn't be `static`?

File: app/src/main/java/com/simplecity/amp_library/ui/screens/album/detail/AlbumDetailFragment.kt [hunk 19]
```diff
@@ -400,6 +403,9 @@ class AlbumDetailFragment :
 
     private var songClickListener: SongView.ClickListener = object : SongView.ClickListener {
         override fun onSongClick(position: Int, songView: SongView) {
+            val uiEventType = UiEventType.PLAY
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/album/detail/AlbumDetailFragment.kt]: I'd get rid of the first line and just pass `UiEventType.PLAY` directly into the below method

File: app/src/main/java/com/simplecity/amp_library/ui/screens/artist/detail/ArtistDetailFragment.kt [hunk 20]
```diff
@@ -151,6 +154,9 @@ class ArtistDetailFragment :
 
     private val songClickListener = object : SongView.ClickListener {
         override fun onSongClick(position: Int, songView: SongView) {
+            val uiEventType = UiEventType.PLAY
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/artist/detail/ArtistDetailFragment.kt]: Same here - just pass `UiEventType.PLAY` directly into method

File: app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java (hunk 18 above)
Comment (barbeau) [reply app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java]: Also, I'd change the method signature to take in the `PlayerEventType`, and pass in `PlayerEventType.PLAY` from where this is called so it doesn't need to be generalized for other event types later.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/folders/FolderFragment.java [hunk 21]
```diff
@@ -478,6 +486,11 @@ public void onFileObjectClick(int position, FolderView folderView) {
         }
     }
 
+    private static void newUiEvent(Song song){
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/folders/FolderFragment.java]: I'd change the method signature to take in the `UiEventType`, and pass in `UiEventType.PLAY` from where this is called so it doesn't need to be generalized for other event types later.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/genre/detail/GenreDetailFragment.kt [hunk 22]
```diff
@@ -447,6 +450,9 @@ class GenreDetailFragment :
     private var songClickListener: SongView.ClickListener = object : SongView.ClickListener {
         override fun onSongClick(position: Int, songView: SongView) {
             if (!contextualToolbarHelper!!.handleClick(songView, Single.just(listOf(songView.song)))) {
+                val uiEventType = UiEventType.PLAY
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/genre/detail/GenreDetailFragment.kt]: Same here - just pass in `UiEventType.PLAY` directly into below method

File: app/src/main/java/com/simplecity/amp_library/ui/screens/miniplayer/MiniPlayerFragment.java [hunk 23]
```diff
@@ -247,4 +261,19 @@ public void showUpgradeDialog() {
             UpgradeDialog.Companion.newInstance().show(getChildFragmentManager());
         }
     };
+
+    @Nullable
+    private Song getSong() {
+        if (MusicServiceConnectionUtils.serviceBinder != null && MusicServiceConnectionUtils.serviceBinder.getService() != null) {
+            return MusicServiceConnectionUtils.serviceBinder.getService().getSong();
+        }
+        return null;
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/miniplayer/MiniPlayerFragment.java:272]: If this can return `null`, please make sure that where this method is called the returned `null` value is handled correctly.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/songs/list/SongListFragment.kt [hunk 24]
```diff
@@ -294,6 +288,9 @@ class SongListFragment :
 
     override fun onSongClick(position: Int, songView: SongView) {
         if (!contextualToolbarHelper!!.handleClick(songView, songView.song)) {
+            val uiEventType = UiEventType.PLAY
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/songs/list/SongListFragment.kt]: Same here - just pass `UiEventType.PLAY` directly into below method.

File: app/src/main/java/com/simplecity/amp_library/ui/screens/suggested/SuggestedFragment.kt [hunk 25]
```diff
@@ -182,6 +185,9 @@ class SuggestedFragment :
     inner class SongClickListener(val songs: List<Song>) : SuggestedSongView.ClickListener {
 
         override fun onSongClick(song: Song, holder: SuggestedSongView.ViewHolder) {
+            val uiEventType = UiEventType.PLAY
```
Comment (barbeau) [inline app/src/main/java/com/simplecity/amp_library/ui/screens/suggested/SuggestedFragment.kt]: Same - just pass into below method directly

File: app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java [hunk 26]
```diff
@@ -0,0 +1,14 @@
+package edu.usf.sas.pal.muser.constants;
+
+public class EventConstants {
+
+    public static String FIREBASE_PLAYER_EVENT_FOLDER = "player-events";
+
+    public static String FIREBASE_UI_EVENT_FOLDER = "ui-events";
+
+    public static String USER_ID = "eventUserId";
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java]: Please change value to `userId`

File: app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+
+    public static String USER_EMAIL = "eventUserEmail";
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java]: Please change value to `userEmail`

File: app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+
+    public static final String RECORD_ID = "eventRecordID";
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/constants/EventConstants.java]: Please change value to `recordId`

File: app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt [hunk 29]
```diff
@@ -1,24 +1,21 @@
 package edu.usf.sas.pal.muser.model
 
 /**
- * A model class that stores the actions performed on the song track. The actions like play, pause,
+ * A model interface that stores the actions performed on the song track. The actions like play, pause,
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/Event.kt]: This documentation should be updated further like:

>A model interface that defines timestamps and song position, along with the song, which can be extended to record various types of events.

File: app/src/main/java/edu/usf/sas/pal/muser/model/PlayerEvent.kt [hunk 30]
```diff
@@ -0,0 +1,18 @@
+package edu.usf.sas.pal.muser.model
+
+/**
+ * A model data class that stores the event occurred while playing a song and implements the
+ * Event interface.
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/PlayerEvent.kt]: Let's change to:

>A model class that implements the Event interface and stores music player events that impact the music the user hears. The timestamp and type of event, song position, and song metadata is captured.

File: app/src/main/java/edu/usf/sas/pal/muser/model/UiEvent.kt [hunk 31]
```diff
@@ -0,0 +1,17 @@
+package edu.usf.sas.pal.muser.model
+
+/**
+ *  A model class that stores different types of events actions performed on the song and implements
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/UiEvent.kt]: Let's change to:

>A model class that implements the Event interface and stores user interface interaction events. The timestamp and type of event, song position, and song metadata is captured.

File: app/src/main/java/edu/usf/sas/pal/muser/model/PlayerEventType.java [hunk 32]
```diff
@@ -0,0 +1,15 @@
+package edu.usf.sas.pal.muser.model;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ *  The enumerator that stores different types of events that occur while playing
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/PlayerEventType.java]: Let's change to:

> Different types of events that occur in the music player that impact the song the user is hearing

File: app/src/main/java/edu/usf/sas/pal/muser/model/UiEventType.java [hunk 33]
```diff
@@ -0,0 +1,16 @@
+package edu.usf.sas.pal.muser.model;
+
+/**
+ * The enumerator that stores different types of action that can be performed by the user
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/model/UiEventType.java]: Let's change to:

>Different types of user interface events triggered by the user

File: app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java [hunk 34]
```diff
@@ -1,30 +1,66 @@
...
-        long elapsedTime =  song.getElapsedTime();
-        return new Event(capturedEvent, currentTimeMS, nanoTime, song.startTime, elapsedTime, songData);
+        long seekPositionMs = getPosition();
+        return new UiEvent(capturedUiAction, currentTimeMS, nanoTime, seekPositionMs,
+                songData);
+    }
+
+    /**
+     * function to get the seek position of the track.
+     * @return current seek position
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/EventUtils.java]: Please update docs to say that `0`, or start of track, is returned in case of error

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+                .addOnCompleteListener(task -> {
+                    if (task.isSuccessful()) {
+                        Log.d(TAG, "UiEvent saved with ID " + documentReference.getId());
+                    } else {
+                        logErrorMessage(task.getException(), "UiEvent Save Failed");
+                    }
+                });
+    }
+
+    private static String getRecordID(){
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: This still needs to be updated:

```suggestion
    private static synchronized String getRecordID(){
```

File: app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java [hunk 36]
```diff
@@ -0,0 +1,115 @@
...
+        long rPrefix = PreferenceUtils.getLong(EventConstants.RECORD_ID, 0);
+        String recordID = rPrefix++ + "-" + UUID.randomUUID().toString();
+        PreferenceUtils.saveLong(EventConstants.RECORD_ID, rPrefix);
+        return recordID;
+    }
+
+    public static void logErrorMessage(Exception e, String message) {
+        if (e != null) {
+            Log.d(TAG, message +
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/FirebaseIOUtils.java]: This and following line can be on a single line.

File: app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java [hunk 37]
```diff
@@ -0,0 +1,47 @@
...
+    public static String getString(String key){
+        return ShuttleApplication.getPrefs().getString(key, null);
+    }
+
+    @TargetApi(9)
+    public static void saveLong(SharedPreferences prefs, String key, long value) {
+        SharedPreferences.Editor edit = prefs.edit();
+        edit.putLong(key, value);
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
```
Comment (barbeau) [inline app/src/main/java/edu/usf/sas/pal/muser/util/PreferenceUtils.java]: Our `minSdkVersion` is definitely more than GINGERBREAD, so you can eliminate this IF/THEN and just have `edit.apply();`

**Review by barbeau — CHANGES_REQUESTED** [review]
@BumbleFlash Thanks for all your work on this! It mostly looks good - a few questions and some documentation suggestions.

File: app/src/main/java/com/simplecity/amp_library/playback/MusicService.java (hunk 15 above)
Comment (BumbleFlash) [reply app/src/main/java/com/simplecity/amp_library/playback/MusicService.java]: Thanks for noticing this. Yeah, it should be PAUSE.

File: app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java (hunk 18 above)
Comment (BumbleFlash) [reply app/src/main/java/com/simplecity/amp_library/playback/QueueManager.java]: Sounds good. Originally I did that but AS was throwing a warning since the only parameter that I was passing is `PlayerEventType.PLAY`.

**Comment by barbeau** [comment]
Thanks @BumbleFlash! This looks good. I'm going to merge `firebase-full` into `usf` as we discussed too, so we can just use `usf` as the base branch for all future pull requests.
